### PR TITLE
[Merged by Bors] - feat: port Data.Option.Defs

### PR DESCRIPTION
**Pull request by javra**

### CODE REVIEW

File: Mathlib/Data/Option/Defs.lean [hunk 1]
```diff
@@ -57,3 +59,71 @@ def melim {α β : Type _} {m : Type _ → Type _} [Monad m] (x : m (Option α))
...
+@[simp]
+theorem mem_to_list {a : α} {o : Option α} : a ∈ toList o ↔ a ∈ o := by
+  cases o <;> simp [toList, eq_comm]
+
+-- lift f
+instance lift_or_get_comm (f : α → α → α) [h : IsCommutative α f] : IsCommutative (Option α) (liftOrGet f) :=
+  ⟨fun a b => by cases a <;> cases b <;> simp [lift_or_get, IsCommutative.comm]⟩
+
+instance lift_or_get_assoc (f : α → α → α) [h : IsAssociative α f] : IsAssociative (Option α) (liftOrGet f) :=
+  ⟨fun a b c => by cases a <;> cases b <;> cases c <;> simp [lift_or_get, IsAssociative.assoc]⟩
```
Comment (javra) [inline Mathlib/Data/Option/Defs.lean:116]: Why does `simp` fail if I instead supply `h.assoc`?

**Comment by javra** [comment]
(This is likely going to break a bit as soon as #493 is merged)

File: Mathlib/Data/Option/Defs.lean (hunk 1 above)
Comment (pechersky) [reply Mathlib/Data/Option/Defs.lean:128]: At this point, you can remove the `[h : ...]` `h` identifer. I also ran into this, you like saw my zulip thread.

File: Mathlib/Data/Option/Defs.lean [hunk 2]
```diff
@@ -57,3 +60,79 @@ def melim {α β : Type _} {m : Type _ → Type _} [Monad m] (x : m (Option α))
...
+  rfl
+
+@[simp]
+theorem mem_toList {a : α} {o : Option α} : a ∈ toList o ↔ a ∈ o := by
+  cases o <;> simp [toList, eq_comm]
+
+#align option.mem_to_list option.mem_toList
+
+-- lift f
+instance lift_or_get_comm (f : α → α → α) [h : IsCommutative α f] :
```
Comment (pechersky) [inline Mathlib/Data/Option/Defs.lean:122]: Based on disagreements on naming conventions, I chose to just make these anonymous instances in my attempt at porting.

**Comment by semorrison** [comment]
> (This is likely going to break a bit as soon as #493 is merged)

Let's merge this one first, since it is lower in the hierarchy.

File: Mathlib/Data/Option/Defs.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Data/Option/Defs.lean:122]: Sadly this is a bit dangerous: this identifier is used in `Algebra.Group.With_One`...!

File: Mathlib/Data/Option/Defs.lean [hunk 3]
```diff
@@ -24,27 +26,28 @@ def lift_or_get (f : α → α → α) : Option α → Option α → Option α
...
+produces a result. -/
 def maybe.{u, v} {m : Type u → Type v} [Monad m] {α : Type u} : Option (m α) → m (Option α)
 | none => pure none
 | some fn => some <$> fn
 
-/-- Map a monadic function `f : α → m β` over an `o : option α`, maybe producing a result. -/
+/-- Map a monadic function `f : α → m β` over an `o : Option α`, maybe producing a result. -/
 def mmap.{u, v, w} {m : Type u → Type v} [Monad m] {α : Type w} {β : Type u} (f : α → m β)
     (o : Option α) : m (Option β) :=
   (o.map f).maybe
```
Comment (semorrison) [inline Mathlib/Data/Option/Defs.lean:53]: This should be called `mapM`, for consistency with monadic functions in Lean 4 core.

File: Mathlib/Data/Option/Defs.lean [hunk 4]
```diff
@@ -57,3 +60,79 @@ def melim {α β : Type _} {m : Type _ → Type _} [Monad m] (x : m (Option α))
 /-- A monadic analogue of `Option.getD`. -/
 def mgetD {α : Type _} {m : Type _ → Type _} [Monad m] (x : m (Option α)) (y : m α) : m α :=
   melim x y pure
```
Comment (semorrison) [inline Mathlib/Data/Option/Defs.lean:62]: And this one probably `getMD` (although I couldn't complain if it were `getDM`...?)

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**